### ISSUE TITLE
Postmortem: Taskcluster Decision Task PR broke master commit checks (multiple times)

### ISSUE THREAD

**Issue by stephenmcgruer** [web-platform-tests/wpt#20523]
**Owner**: @stephenmcgruer 
**Postmortem Created**: 2019-11-28 21:45 EST
**Status**: In Review
**Issue**: No specific issue for landing the Taskcluster Decision Graph change exists. See [here](https://github.com/web-platform-tests/wpt/issues/20256#issuecomment-557289766).

**Impact**: Medium. Approximately 14 (4 for the first landing, 10 for the second) commits to master did not run Taskcluster successfully. This means that experimental Chrome and Firefox results for those commits are not available on wpt.fyi ([example](https://wpt.fyi/api/runs?sha=60d8480d2c1e5a54982ac9b1223375bc407bc40f)). The change also broke the epoch branches responsible for producing stable runs, and we lose a few days of stable results.

**Root Cause**: A planned change from a single status Taskcluster yaml file to a 'decision task' approach had a number of bugs and unintended side effects, which were only discovered as it rolled out.

**Timeline**
- 2019-11-22 11:47 EST: The first attempt at creating a decision task for Taskcluster [is merged](https://github.com/web-platform-tests/wpt/pull/19401#event-2823887393).
- 2019-11-22 11:48 EST: Subsequent commits (from other PRs) land on master and begin to fail ([example](https://community-tc.services.mozilla.com/tasks/groups/e914EZsUTwWxi6yCyAAeiQ)).
- **FIRST OUTAGE BEGINS**
- 2019-11-22 14:33 EST: Having identified [three known bugs](https://github.com/web-platform-tests/wpt/pull/20407#issuecomment-557665261), @Hexcles [reverts](https://github.com/web-platform-tests/wpt/pull/20407#event-2824383081) the Taskcluster decision task CL.
- **FIRST OUTAGE ENDS**
- 2019-11-22 14:33 EST: A [reland PR](https://github.com/web-platform-tests/wpt/pull/20408) is created, with the intention to address the known issues before relanding.
- 2019-11-26 16:24 EST: Having gone through some rounds of review, part of the change is [pulled into a separate PR](https://github.com/web-platform-tests/wpt/pull/20463) and landed without issue.
- 2019-11-28 09:20 EST: The reland PR [is merged](https://github.com/web-platform-tests/wpt/pull/20408#event-2839120313).
- 2019-11-28 09:21 EST: Subsequent commits (from other PRs) land on master and begin to fail ([example](https://community-tc.services.mozilla.com/tasks/groups/RB8weoWURF-ab971eJ63ow)). This is noticed by the Taskcluster decision task PR author, who begins working on a fix.
- **SECOND OUTAGE BEGINS**
- 2019-11-28 10:32 EST: A fix for the believed issue [is landed](https://github.com/web-platform-tests/wpt/pull/20507#event-2839334694). Unfortunately, it does not address the problems, and it appears (??) that no-one notices that master is still failing.
- 2019-11-28 17:16 EST: Whilst working on an [incidental PR](https://github.com/web-platform-tests/wpt/pull/20518), @stephenmcgruer notices that the Taskcluster run for it has hung. Away from his laptop, he requests (via internal chat) that members of the Chromium Ecosystem Infra team investigate whether this is widespread.
- 2019-11-28 17:54 EST: @KyleJu confirms that every commit since the decision task PR landed has failed, for reasons similar to the first time.
- 2019-11-28 18:24 EST: Back at a laptop, @stephenmcgruer puts together a [revert PR](https://github.com/web-platform-tests/wpt/pull/20520) of the decision task PR plus related changes since.
- 2019-11-28 18:55 EST: The revert [is merged](https://github.com/web-platform-tests/wpt/pull/20520#event-2840041855).
- **SECOND OUTAGE ENDS**
- 2019-11-28 20:03 EST: @stephenmcgruer confirms that master has gone green again.
- 2019-11-29 08:46 EST: Having found the fix for the latest problem + discussed on IRC, @jgraham and @stephenmcgruer [agree to try once more to land the decision task PR](https://github.com/web-platform-tests/wpt/pull/20528#issuecomment-559794884).
- 2019-11-29 08:56 EST: @jgraham and @stephenmcgruer observe that master appears to be running ok (no failure yet), but [strange errors](https://github.com/web-platform-tests/wpt/commit/8e175af1b2917b10942077d4609947937db71b92#commitcomment-36194023) from Taskcluster are seen on the individual commits.
- 2019-11-29 09:23 EST:  Working theory is that Taskcluster is trying to respond to events other than pushes and failing, but there appears to be no way to tell. An attempt is made to contact Taskcluster engineers on #taskcluster to get logs.
- 2019-11-29 09:47 EST: Master commit for Taskcluster decision task [successfully passes](https://github.com/web-platform-tests/wpt/commit/8e175af1b2917b10942077d4609947937db71b92). \o/
- 2019-11-29 09:52 EST: @stephenmcgruer realizes that the task names changed, so wpt.fyi won't be able to parse them. @jgraham starts working on PR to change the names back.
- 2019-11-29 10:20 EST: Task name fix [merged](https://github.com/web-platform-tests/wpt/pull/20531#event-2841733422).
- 2019-11-29 10:23 EST: @jgraham notices that PR checks are failing on chrome-stability specifically.
- 2019-11-29 11:00 EST: @stephenmcgruer finds the cause; xvfb isn't enabled for chrome-stability after the decision task change. A [fix is uploaded](https://github.com/web-platform-tests/wpt/pull/20533).
- 2019-11-29 11:16 EST: Fix for chrome-stability [lands](https://github.com/web-platform-tests/wpt/pull/20533#event-2841862687).
- 2019-11-29 11:30 EST: @stephenmcgruer confirms the fix appears to have worked.
- 2019-12-02 07:18 EST: @rakuco discovers that stable runs are missing for [Chrome, Firefox](https://github.com/web-platform-tests/wpt/issues/20551) and experimental runs for [WebKitGTK](https://github.com/web-platform-tests/wpt/issues/20552). @stephenmcgruer begins to investigate.
- 2019-12-02 09:23 EST: @stephenmcgruer discovers an [overlooked error](https://github.com/web-platform-tests/wpt/issues/20551#issuecomment-560417230) in an epochs commit, which confirms the outage was caused by a bug in the Taskcluster decision task change.
- 2019-12-02 12:53 EST: A [fix is merged](https://github.com/web-platform-tests/wpt/pull/20561#event-2846656291)
- 2019-12-03 05:00 EST: Fix is confirmed to work, the runs are back.

**Lessons Learnt**

*Things that went well*

- N/A, it seems :(

*Things that went poorly*

- It is difficult to test changes to the Taskcluster configurations properly. One can push to a trigger branch and the PR author did so, but missed the timeouts as we saw some chunks go green ([logs](https://community-tc.services.mozilla.com/tasks/groups/S7-yip4XTXKDSAPjpM0_tQ)). However, it really isn't possible to test the impact of a TaskCluster configuration change on PR checks.
- In the second outage, @stephenmcgruer had been asked to monitor the Taskcluster decision task post-landing. However he was unaware that the problems last time had been on *commited* PRs, and so had been looking at in-flight PRs and seeing no issues.
- It appears that no-one involved received or noticed notifications (email or github) that master commits were failing checks in either case.

*Where we got lucky*

- In the second outage, @stephenmcgruer just happened to notice that his in-flight PR had hung; otherwise the outage would have gone unnoticed until at least the next day.
- In the second outage, both @stephenmcgruer and @KyleJu happened to be available after working hours to debug and revert.

**Action Items**

- [ ] Make sure that notifications are sent if master commits are failing the push checks _(type=detect, owner=???)_
- [ ] Add more tests to https://github.com/web-platform-tests/wpt/blob/master/tools/ci/tc/tests/test_valid.py _(type=prevent, owner=???)_

**Comment by Hexcles** [comment]
I'm proposing an action item: more tests for https://github.com/web-platform-tests/wpt/blob/master/tools/ci/tc/tests/test_valid.py

https://github.com/web-platform-tests/wpt/pull/20533#issuecomment-559873813

>  I think maybe we should really check in the full expected JSONs and deep compare scheduled.

**Comment by stephenmcgruer** [comment]
@jgraham @Hexcles - I noticed that I had just let this postmortem sit uncompleted from November :(. I tidied it up a little and added a few AIs, but I'm not sure whether we want to action them or not. Lmk what you think (and if you have any other suggestions for the postmortem)

**Comment by foolip** [comment]
Over a year later, I'm gonna go ahead and close this :)
